File: practice/Tower of Tablets/TowerofTablets.py
def reorder_tablets(old_list, position, new_list):
    if not old_list:
        return True

    for i in range(len(old_list)):
        current_tablet = old_list[i]
        last_tablet = new_list[-1]

        # Check if the glyph at the given position matches
        if current_tablet[position] == last_tablet[position]:
            # Create new lists for backtracking
            new_old_list = old_list[:i] + old_list[i + 1 :]
            new_new_list = new_list + [current_tablet]

            # Try both positions recursively
            if reorder_tablets(new_old_list, 1 - position, new_new_list):
                new_list.clear()
                new_list.extend(new_new_list)
                return True

    return False


def main():
    n = int(input())  # Number of tablets
    tablets = []

    # Read the glyphs of tablets
    for _ in range(n):
        glyph = input().strip()
        tablets.append(glyph)

    is_possible = False

    for i in range(len(tablets)):
        old_list = tablets[:i] + tablets[i + 1 :]
        new_list = [tablets[i]]

        if reorder_tablets(old_list, 0, new_list) or reorder_tablets(
            old_list, 1, new_list
        ):
            is_possible = True
            for tablet in new_list:
                print(tablet)
            break

    if not is_possible:
        print(-1)


if __name__ == "__main__":
    main()

File: practice/Tower of Tablets/TowerofTablets.java
import java.util.ArrayList;
import java.util.Scanner;

public class TowerofTablets {
    // TIME LIMIT EXCEEDED
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // the number of tablets
        int n = scanner.nextInt();
        scanner.nextLine();

        ArrayList<String> tablets = new ArrayList<>();

        // the glyphs of tablets
        for (int i = 0; i < n; i++) {
            String glyph = scanner.nextLine();
            tablets.add(glyph);
        }

        boolean isPossible = false;

        for (int i = 0; i < tablets.size(); i++) {
            ArrayList<String> oldList = new ArrayList<>(tablets);
            ArrayList<String> newList = new ArrayList<>();
            newList.add(tablets.get(i));
            oldList.remove(i);

            if (reorderTablets(oldList, 0, newList) || reorderTablets(oldList, 1, newList)) {
                isPossible = true;
                for (String tablet : newList) {
                    System.out.println(tablet);
                }
                break;
            }
        }

        if (!isPossible) {
            System.out.println(-1);
        }

        scanner.close();
    }

    public static boolean reorderTablets(ArrayList<String> oldList, int position, ArrayList<String> newList) {
        if (oldList.isEmpty()) {
            return true;
        }

        for (int i = 0; i < oldList.size(); i++) {
            String currentTablet = oldList.get(i);
            String lastTablet = newList.get(newList.size() - 1);

            // Check if the glyph at the given position matches
            if (currentTablet.charAt(position) == lastTablet.charAt(position)) {
                ArrayList<String> newOldList = new ArrayList<>(oldList);
                ArrayList<String> newNewList = new ArrayList<>(newList);

                newNewList.add(currentTablet);
                newOldList.remove(i);

                // Try both positions recursively
                if (reorderTablets(newOldList, 1 - position, newNewList)) {
                    newList.clear();
                    newList.addAll(newNewList);
                    return true;
                }
            }
        }

        return false;
    }
}